src/test/scala/io/iohk/ethereum/blockchain/sync/FastSyncStateHandlerSpec.scala

```diff
@@ -0,0 +1,425 @@
+
+
+    val expectedState = startState.copy(
+        pendingMptNodes = Seq(StateMptNodeHash(correctNewTargetBlock.stateRoot)),
+        updatingTargetBlock = false
+    )
+
+    val (comm, state1) = stateHandler.handleNewTargetBlock(ImportedLastBlock, correctNewTargetBlock)(startState)
+    state1 shouldEqual expectedState
+    comm match {
+      case BackToSyncing => succeed
+      case _ => fail
+    }
+  }
+
+  it should "properly handle new target block, and switch to higher block" in new TestSetup {
+    val targetBlockNumber = 5000
+    val targetBlock = defaultTargetBlockHeader.copy(number = targetBlockNumber)
+    val safe = targetBlockNumber + syncConfig.fastSyncBlockValidationX
+
+
+    val startState = SyncState(
+      targetBlock,
+      safe,
+      bestBlockHeaderNumber = safe,
+      updatingTargetBlock = true
+    )
+
+    val correctNewTargetBlock = targetBlock.copy(number = targetBlockNumber + syncConfig.maxTargetDifference + 1)
+
+
+    val expectedState = startState.copy(
+      targetBlock = correctNewTargetBlock,
+      safeDownloadTarget = correctNewTargetBlock.number + syncConfig.fastSyncBlockValidationX,
+      updatingTargetBlock = false
+    )
+
+    val (comm, state1) = stateHandler.handleNewTargetBlock(ImportedLastBlock, correctNewTargetBlock)(startState)
+    state1 shouldEqual expectedState
+    comm match {
+      case BackToSyncing => succeed
+      case _ => fail
+    }
+  }
+
+  it should "re-ask for new target block, after receiving target block which has lower number" in new TestSetup {
+    val targetBlockNumber = 5000
+    val targetBlock = defaultTargetBlockHeader.copy(number = targetBlockNumber)
+    val safe = targetBlockNumber + syncConfig.fastSyncBlockValidationX
+
+
+    val startState = SyncState(
+      targetBlock,
+      safe,
+      bestBlockHeaderNumber = safe,
+      updatingTargetBlock = true
+    )
+
+    val correctNewTargetBlock = targetBlock.copy(number = targetBlockNumber - 1)
+
+
+    val expectedState = startState.copy(
+      targetBlockUpdateFailures = 1,
+      updatingTargetBlock = true
+    )
+
+    val (comm, state1) = stateHandler.handleNewTargetBlock(ImportedLastBlock, correctNewTargetBlock)(startState)
+    state1 shouldEqual expectedState
+    comm match {
+      case AskForNewTarget(_) => succeed
+      case _ => fail
+    }
+  }
+
+  it should "properly handle new target block update, after validation failure" in new TestSetup {
+    val targetBlockNumber = 5000
+    val targetBlock = defaultTargetBlockHeader.copy(number = targetBlockNumber)
+    val safe = targetBlockNumber + syncConfig.fastSyncBlockValidationX
+
+
+    val startState = SyncState(
+      targetBlock,
+      safe,
+      bestBlockHeaderNumber = safe,
+      updatingTargetBlock = true
+    )
+
+    val correctNewTargetBlock = targetBlock.copy(number = targetBlockNumber + syncConfig.maxTargetDifference + 1)
+
+
```

Comment:
maybe one enter should work, same as in 163-164, 192-193, 202-203, 222-223, 232-233, 251-252